src/mlpack/base.hpp

```diff
@@ -0,0 +1,124 @@
+/**
+ * @file base.hpp
+ *
+ * The most basic core includes that mlpack expects; standard C++ includes and
+ * Armadillo only
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_BASE_HPP
+#define MLPACK_BASE_HPP
+
+// First, check if Armadillo was included before, warning if so.
+#ifdef ARMA_INCLUDES
+#pragma message "Armadillo was included before mlpack; this can sometimes cause\
+ problems.  It should only be necessary to include <mlpack/core.hpp> and not \
+<armadillo>."
+#endif
+
+// Defining _USE_MATH_DEFINES should set M_PI.
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+// Next, standard includes.
+#include <cctype>
+#include <cfloat>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+#include <numeric>
+#include <vector>
+#include <queue>
+
+// But if it's not defined, we'll do it.
+#ifndef M_PI
+  #define M_PI 3.141592653589793238462643383279
+#endif
+
+// MLPACK_COUT_STREAM is used to change the default stream for printing
+// purpose.
+#if !defined(MLPACK_COUT_STREAM)
+ #define MLPACK_COUT_STREAM std::cout
+#endif
+
+// MLPACK_CERR_STREAM is used to change the stream for printing warnings
+// and errors.
+#if !defined(MLPACK_CERR_STREAM)
+ #define MLPACK_CERR_STREAM std::cerr
+#endif
+
+// Give ourselves a nice way to force functions to be inline if we need.
+#define force_inline
+#if defined(__GNUG__) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __attribute__((always_inline))
+#elif defined(_MSC_VER) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __forceinline
+#endif
+
+// Backport this functionality from C++14, if it doesn't exist.
+#if __cplusplus <= 201103L
+#if !defined(_MSC_VER) || _MSC_VER <= 1800
+namespace std {
+
+template<bool B, class T = void>
+using enable_if_t = typename enable_if<B, T>::type;
+
+}
+#endif
+#endif
+
+// Backport std::any from C+17 to C++11 to replace boost::any.
+// Use mnmlstc backport implementation only if compiler does not
+// support C++17.
+#if __cplusplus < 201703L
+  #include <mlpack/core/std_backport/any.hpp>
+  #include <mlpack/core/std_backport/string_view.hpp>
+  #define ANY core::v2::any
+  #define ANY_CAST core::v2::any_cast
+  #define STRING_VIEW core::v2::string_view
```

Comment:
Suggest to prefix all these `#defines` with `MLPACK_` to avoid potential clashes with names in user code.  For example, change `ANY` to `MLPACK_ANY`.  This is especially applicable to `ANY`, since its dangerously only a 3 letter macro.

Note that the `#define` macros apply to all strings in source code, so if a user creates a function called `ANY(int x)`, it will be replaced with `core::v2::any(int x)`, which doesn't make sense and is clearly undesirable.  It will blow up with a confusing error message.

Using the `MLPACK_` prefix will considerably reduce the likelihood of inadvertent clashes.

(cf. MS really screwed up when they shortsightedly defined `min` and `max` as macros in their silly windoze headers, causing no end to grief; GNU glibc also screwed up by having `major` and `minor` defined as macros -- see `sys/sysmacros.h`: https://sourceware.org/bugzilla/show_bug.cgi?id=19239).

Reply:
Yeah, agreed; done in c02eba2ec9.